libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/NavigationTelemetry.kt

```diff
@@ -0,0 +1,562 @@
+package com.mapbox.services.android.navigation.v5.internal.navigation
+
+import android.annotation.SuppressLint
+import android.app.Application
+import android.content.Context
+import android.location.Location
+import com.mapbox.android.core.location.LocationEngine
+import com.mapbox.android.telemetry.TelemetryUtils
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.core.constants.Constants
+import com.mapbox.geojson.utils.PolylineUtils
+import com.mapbox.services.android.navigation.BuildConfig
+import com.mapbox.services.android.navigation.v5.internal.exception.NavigationException
+import com.mapbox.services.android.navigation.v5.internal.location.MetricsLocation
+import com.mapbox.services.android.navigation.v5.internal.navigation.metrics.FeedbackEvent
+import com.mapbox.services.android.navigation.v5.internal.navigation.metrics.NavigationMetricListener
+import com.mapbox.services.android.navigation.v5.internal.navigation.metrics.RerouteEvent
+import com.mapbox.services.android.navigation.v5.internal.navigation.metrics.SessionState
+import com.mapbox.services.android.navigation.v5.internal.navigation.routeprogress.MetricsRouteProgress
+import com.mapbox.services.android.navigation.v5.internal.utils.RingBuffer
+import com.mapbox.services.android.navigation.v5.navigation.MapboxNavigation
+import com.mapbox.services.android.navigation.v5.navigation.MapboxNavigationOptions
+import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgress
+import java.util.ArrayList
+import java.util.Date
+import java.util.Locale
+import java.util.concurrent.TimeUnit
+
+@SuppressLint("StaticFieldLeak")
+internal object NavigationTelemetry : NavigationMetricListener {
+    private const val MAPBOX_NAVIGATION_SDK_IDENTIFIER = "mapbox-navigation-android"
+    private const val MAPBOX_NAVIGATION_UI_SDK_IDENTIFIER = "mapbox-navigation-ui-android"
+    private const val MOCK_PROVIDER =
+        "com.mapbox.services.android.navigation.v5.location.replay.ReplayRouteLocationEngine"
+    private const val TWENTY_SECOND_INTERVAL = 20
+    private const val LOCATION_BUFFER_MAX_SIZE = 40
+
+    private lateinit var context: Context
+    private lateinit var eventDispatcher: NavigationEventDispatcher
+    private lateinit var departEventFactory: DepartEventFactory
+
+    private val queuedRerouteEvents = ArrayList<RerouteEvent>()
+    private val queuedFeedbackEvents = ArrayList<FeedbackEvent>()
+    private val locationBuffer: RingBuffer<Location> = RingBuffer(LOCATION_BUFFER_MAX_SIZE)
+    private var metricProgress: MetricsRouteProgress = MetricsRouteProgress(null)
+    private var navigationSessionState: SessionState = SessionState.builder().build()
+    private var metricLocation: MetricsLocation = MetricsLocation(null)
+    private val gpsEventFactory = InitialGpsEventFactory()
+    private var isOffRoute: Boolean = false
+    private var isInitialized = false
+
+    private var lifecycleMonitor: NavigationLifecycleMonitor? = null
```

Comment:
Could this be `lifecycleMonitor` be non-nullable using `lateinit`?

Reply:
It's fine as `NavigationTelemetry` could be alive greater time than `NavigationService` and we want to reset the values.